text/0113-secure-script-container.md

```diff
@@ -0,0 +1,178 @@
+
+## Motivation
+
+Web component based applications need the ability to integrate with third party libraries. Because
+web components rely on the Shadow DOM for encapsulation, this development paradigm does not work
+with libraries who are expecting to interact with the application globally. Before there was
+component encapsulation via Shadow DOM it was possible for library authors to:
+
+- Listen for events and know who fired them
+- Define global styling rules that would get applied on every element
+- Query the entire document for any element on the page
+
+These libraries are typically added as global scripts to the document root rather than deeper 
+integrations into the components themselves. When every component in the application has a ShadowRoot attached to them, these libraries will not work as expected.
+
+Some examples of these libraries are:
+
+- **Analytics tools**: Google Tag Manager / Google Analytics, Pendo
+- **Personalization platforms**: Optimizely, Evergage, Google Optimize
+- **Instrumentation tools**: New Relic, Sentry.io
+
+The goal of this proposal is to provide a way for application developers to continue to use their existing integrations until library authors support Shadow DOM integration.
+
+## Design
+
+A new element called `<virtual-script>` encapsulates the various mechanisms to evaluate the scripts in a virtual environment, which solves the problem when using synthetic shadow and subsequently, when using the native Shadow DOM. We can implement this solution in two phases:
+
+### Phase I: Synthetic Shadow
+
+Many LWC applications use the [synthetic-shadow polyfill](https://github.com/salesforce/lwc/tree/master/packages/%40lwc/synthetic-shadow/). This polyfill emulates the native Shadow DOM behavior while still allowing global styles to cascade into the shadow trees. Synthetic Shadow patches many of the DOM APIs that allow you to interact and traverse the DOM tree, (e.g.: `document.querySelector` and `document.querySelectorAll`), these APIs will prevent developers from penetrating the Shadow DOM boundaries defined by their LWC components. In this phase, we want to solve the LWC + Synthetic Shadow scenario.
+
+To ensure we can solve the problem while continuing to guarantee the benefits that are provided by the synthetic-shadow polyfill for all other components running on the main window, we are currently proposing the following characteristics for the virtual-script design:
+
+* Create a JS sandbox before the Synthetic Shadow Polyfill is evaluated.
+  * The sandbox is created using a same-domain iframe.
+
+* Use this JS sandbox to virtualize the DOM APIs in order to prevent these APIs to be subject to the Synthetic Shadow Polyfill patches:
+  * This can be done by caching references to the to-be-patched APIs inside the sandbox before the synthetic shadow is evaluated.
+
+* Evaluate one or more third party libraries inside this sandbox
```

Comment:
I think we may need a few more details here about how exactly this code gets evaluated. In the POC we are creating a `script` element and appending it to the `magicBody`. Does that actually get added to the `virtual-script`'s Shadow DOM? What if it fails?